### PR TITLE
Ak #1290 view visibility dropinui

### PR DESCRIPTION
**Pull request by abhishek1508**
### Description

Fixes https://github.com/mapbox/navigation-sdks/issues/1290

### CODE REVIEW

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/state/ManeuverState.kt
Comment (abhishek1508) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/state/ManeuverState.kt:22]: I had to create a new `ManeuverErrorFactory` that would allow to construct this object. Since we have marked `*Factory` with @ExperimentalMapboxNavigationAPI, this has taken affect upstream to `MainActivity`. We need to make a decision if it's ok to add this tag to an Activity, or do we want to remove this tag from any `*Factory` objects that we create.
/cc @LukasPaczos

Comment (LukasPaczos) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/state/ManeuverState.kt:22]: @abhishek1508 we don't have to mark all classes that use the new experimental factory with `@ExperimentalMapboxNavigationAPI` (which then goes up and up the usage tree), we just need to tag the specific internal usages with `@OptIn(ExperimentalMapboxNavigationAPI::class)` which does not bubble up.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/contract/ManeuversContract.kt
Comment (LukasPaczos) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/contract/ManeuversContract.kt:13]: This shouldn't be an action that developers have access to. This is an internal logic (and method of transferring data), so this and all other classes below should all be internal.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt
Comment (LukasPaczos) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt:289]: We'll need to have a more extensive logic here and only run an update if the view actually is an instance of `MapboxSpeedLimitView` (same for all other views). If someone would use a custom view for `speedLimitView`, we'd crash here.

If it's not our view, we shouldn't even subscribe to the view model for any updates. We need a controlling unit that understands the configuration and which views are ours and which aren't so that we can skip them.

Comment (LukasPaczos) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt:289]: Let's also try to move it out of the `NavigationView` class so that it doesn't grow.

Comment (abhishek1508) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt:289]: > If it's not our view, we shouldn't even subscribe to the view model for any updates. We need a controlling unit that understands the configuration and which views are ours and which aren't so that we can skip them.

I agree with this totally and was already planning to cut a ticket on this task. I will remove this line here in this PR as this might need some thoughts on how to do even though I have something in mind. But, let's not block this PR and introduce a proper system for it separately.

> Let's also try to move it out of the NavigationView class so that it doesn't grow.

Do you mean all `*.collect { }` calls as there are several

/cc @LukasPaczos

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt
Comment (cafesilencio) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt:70]: I don't think the NavigationView should be creating all of these dependencies for other classes like view models. Just because we're not using a dependency injection framework doesn't mean the NavigationView should take on this role. The providing of dependencies should be abstracted from the view models as well as the NavigationView. No need to hold up this PR for this issue. I'll propose an alternative to this separately.

**Comment by cafesilencio** [comment]
PR https://github.com/mapbox/mapbox-navigation-android/pull/5086 is landing momentarily. After it's merged please make adjustments to `NavigationView` so that rather than public functions, the calls are exposed through the `MapboxNavigationViewApi`.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt
Comment (cafesilencio) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt:70]: In fact the view model factories would be a good place to move these to.

Comment (abhishek1508) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt:70]: Yep I agree and I had the same intention to improve this later and focus on the item relevant to this PR. Thanks!

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt
Comment (cafesilencio) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt:143]: These should be immutable IMO.

**Comment by abhishek1508** [comment]
Hey @LukasPaczos 

Based on our discussion, I have updated the code in this PR to reflect 2 changes:
- Viewmodels are now responsible to read the last value using the flow.
- I have tried to move the `.*collect` calls out of the `NavigationView` as it is growing rapidly. For that effect, instead of arranging the package structure horizontally (contract, viewmodel, state, processor), I have now arranged the package structure vertically based on features(maneuver, recenter, routeoverview, sound, tripprogress, speedlimit). The concept is that DropIn is made up of several standalone components and so I have introduced a `.*Component` class for every feature. For ex:
```
class ManeuverComponent(
    val view: View,
    val container: FrameLayout,
    val viewModel: ManeuverViewModel
)
```
This has allowed me to trigger all actions in the `*.Component` as well as hook up all the state observer changes in the components. One important place where I want a second thought is would this `maneuverComponent` be garbage collected when `NavigationView` is destroyed on configuration change.

**Comment by abhishek1508** [comment]
@LukasPaczos This is ready for another round of review. Given the time constraint and the fact that this PR is growing fast and needs to merge asap I have created this [ticket](https://github.com/mapbox/navigation-sdks/issues/1316) to track adding the missing unit tests. I would also not worry about the `internalObservers` I have added at the moment. We can improve it in further iterations.

File: gradle/jacoco.gradle
Comment (LukasPaczos) [inline gradle/jacoco.gradle:10]: Why is this needed here? I imagine you were running into some problems with unit tests and Java 11?

File: examples/src/main/java/com/mapbox/navigation/examples/core/MapboxDropInActivity.kt
Comment (LukasPaczos) [inline examples/src/main/java/com/mapbox/navigation/examples/core/MapboxDropInActivity.kt:13]: This can be removed.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/MapboxNavigationViewApi.kt
Comment (LukasPaczos) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/MapboxNavigationViewApi.kt:38]: This can be removed or changed to `@OptIn(ExperimentalCoroutinesApi::class)`.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt
Comment (LukasPaczos) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt:80]: The `@ExperimentalCoroutinesApi` should be changed to `@OptIn(ExperimentalCoroutinesApi::class)`.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt
Comment (LukasPaczos) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt:109]: Why do we not have a `MapboxRouteLineComponent`? Not a blocker, just wondering if this is intentional or planned as follow up.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt
Comment (LukasPaczos) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt:436]:
```suggestion
                    when (uiComponent) {
                        is RouteProgressObserver -> {
                            uiComponent.onRouteProgressChanged(routeProgress)
                        }
```

Same below where appropriate.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt
Comment (LukasPaczos) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt:459]: We should have a `component` to manage the puck and wire these operations through that `component`.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt
Comment (LukasPaczos) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt:470]: Same as above, we should have a component for the camera. Can be added separately though.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/maneuver/ManeuverViewModel.kt
Comment (LukasPaczos) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/maneuver/ManeuverViewModel.kt:49]: We don't need a distinct "update visibility" action. Visibility is based on state and data which are both separate actions.

Comment (LukasPaczos) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/maneuver/ManeuverViewModel.kt:49]: Same everywhere below.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/maneuver/ManeuverState.kt
Comment (LukasPaczos) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/maneuver/ManeuverState.kt:14]: Here only `isVisible` and `maneuver` are actual result values which are then used to update the UI. `navigationState` is a data point out of which the result is computed.

We might want to separate these to have 2 object, one object that holds `isVisible` and `maneuver` state, and the other that holds previous input values for reprocessing when values are updated.

Comment (LukasPaczos) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/maneuver/ManeuverState.kt:14]: Same in all other components.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt
Comment (abhishek1508) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt:470]: I agree and I think it would make sense to address this change and above in your camera PR.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/maneuver/ManeuverViewModel.kt
Comment (abhishek1508) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/maneuver/ManeuverViewModel.kt:49]: I had it exactly like that first where a change in `NavigationState` would also update the visibility. Although it makes complete sense, I think we loose some control here and out of nowhere without any intention changing the state changes the visibility as well. I agree it is probably verbose, but it makes more sense to me trigger the action to update visibility rather than some other action that updates a different part of the state. If we would have to travel down the history and track actions, it would become difficult to figure out which action changed the visibility. Also, if in the future we were to change the fact that visibility is not dictated by navigation state then not having a separate action for it will break.

File: gradle/jacoco.gradle
Comment (abhishek1508) [reply gradle/jacoco.gradle:10]: Yeah let me try a different Java version and see if that fixes the problem

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt
Comment (abhishek1508) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt:109]: The route line just got merged yesterday, so probably was missed.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/maneuver/ManeuverState.kt
Comment (abhishek1508) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/maneuver/ManeuverState.kt:14]: I am wondering what is the advantage of separating then out into separate components?

File: gradle/jacoco.gradle
Comment (LukasPaczos) [reply gradle/jacoco.gradle:10]: Vitaly mentioned that he'll help to drive https://github.com/mapbox/mapbox-navigation-android/pull/4454 to the finish line so let's remove this change from here for now.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt
Comment (LukasPaczos) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt:109]: Okay, feel free to add here or cut a ticket :+1:

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt
Comment (LukasPaczos) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt:470]: :ok_hand:

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/maneuver/ManeuverViewModel.kt
Comment (LukasPaczos) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/maneuver/ManeuverViewModel.kt:49]: Why would a change in state ever not change the visibility, or how would we change the state unintentionally? It seems redundant to me that we need to call a visibility change action, it's more of a source of bugs if we _forget_ to call it after the state change. Following the example, we'd have to have an empty action for each operation (maneuver update, distance remaining update, etc).

I still think that we can remove `UpdateVisibility` action and only reduce the available data into a result which is applied, without empty triggers like this.

> Also, if in the future we were to change the fact that visibility is not dictated by navigation state then not having a separate action for it will break.

That would be a breaking change for the entire `component` anyway so that might also be a better point in time to add the necessary action instead of introducing it now without a clear use-case.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/maneuver/ManeuverState.kt
Comment (LukasPaczos) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/maneuver/ManeuverState.kt:14]: Readability in my opinion. It wasn't immediately clear to me why do we store the `navigationState` here since it is not used for any view operations directly.

I understand that it's here to be able to reduce the state when new route progress comes in but it might be worth decoupling the actual state that views feed off of from the internal view model's state.

Comment (LukasPaczos) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/maneuver/ManeuverState.kt:14]: Not a very strong point here though, just my small observation. We can address this later as we add more state properties and see what fits where.

Comment (abhishek1508) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/maneuver/ManeuverState.kt:14]: Yep. I believe we are still not clear on what other data points would be required for our components. It would probably make more sense to update this if required once we have a good amount of functionality ready so then we have more clear understanding. For ex: we are at the moment not tracking any UI state like storing the recycler view scroll position that would help to show the recycler view from the position you left it.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/maneuver/ManeuverViewModel.kt
Comment (abhishek1508) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/maneuver/ManeuverViewModel.kt:49]: > it's more of a source of bugs if we forget to call it after the state change. Following the example, we'd have to have an empty action for each operation (maneuver update, distance remaining update, etc).

Agreed with this. We could probably remove the empty `UpdateVisibility`, but we would later have to add a new action UpdateVisibility(show = true/false) if an end user wants to show a view which is hidden or vice versa
cc @LukasPaczos

Comment (LukasPaczos) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/maneuver/ManeuverViewModel.kt:49]: > we would later have to add a new action UpdateVisibility(show = true/false) if an end user wants to show a view which is hidden or vice versa

I still hope this will be controlled by the state variable which would have an understanding of which component should be visible rather than components hardcoding if they are visible for given state (as things currently are).

Comment (abhishek1508) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/maneuver/ManeuverViewModel.kt:49]: That would be an ideal scenario. In reality the visibility of a view is controlled by many different factors here and is not the same for every view. `Recenter` uses `navigationState` and `cameraState` whereas `Maneuver` uses `navigationState` and `maneuverValue`. It is probably too soon to say if there are any other states/data point that dictate the visibility. Also, it's probably difficult to predict if an end user would just want to show the view irrespective of how it is dictated by the state/data. 

However, we can remove it for now and cross that bridge when we come to it.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt
Comment (abhishek1508) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt]: Added `RouteLineUIComponent` and `RouteArrowUIComponent`
@LukasPaczos

**Review by LukasPaczos — APPROVED** [review]
Great work!

Let's ship and follow up on the remaining changes that we've been tracking separately.

**Pull request merged**
